src/Compilers/Server/VBCSCompiler/BuildServerController.cs

```diff
@@ -39,30 +49,78 @@ internal int Run(string[] args)
                 : RunServer(pipeName, tempPath, cancellationToken: cancellationTokenSource.Token);
         }
 
-        protected internal abstract TimeSpan? GetKeepAliveTimeout();
-
-        protected abstract string GetDefaultPipeName();
-
-        protected abstract IClientConnectionHost CreateClientConnectionHost(string pipeName);
-
-        protected abstract Task<Stream> ConnectForShutdownAsync(string pipeName, int timeout);
+        internal TimeSpan? GetKeepAliveTimeout()
+        {
+            try
+            {
+                int keepAliveValue;
+                string keepAliveStr = _appSettings[KeepAliveSettingName];
+                if (int.TryParse(keepAliveStr, NumberStyles.Integer, CultureInfo.InvariantCulture, out keepAliveValue) &&
+                    keepAliveValue >= 0)
+                {
+                    if (keepAliveValue == 0)
+                    {
+                        // This is a one time server entry.
+                        return null;
+                    }
+                    else
+                    {
+                        return TimeSpan.FromSeconds(keepAliveValue);
+                    }
+                }
+                else
```

Comment:
I get that we're trying not to change implementation details, but it feels weird that we're not at least logging that a setting was bad.